COMPACT_DENYLIST_FORMAT.md

```diff
@@ -0,0 +1,458 @@
+    and should not have a trailing `/`.
+
+The latter form allows blocking by double-hash using any hashing function of
+choice. Implementations will have to hash requests using all the hashing functions
+used in the denylist, so we recommend sticking to one.
+
+Conveniently, the latter form allows using a b58-encoded sha256 multihashes
+(usual form of CIDv0 - `Qmxxx...`), so that double-hashes can be like:
+
+```
+$ printf "QmecDgNqCRirkc3Cjz9eoRBNwXGckJ9WvTdmY16HP88768/my/path" | ipfs add --raw-leaves --only-hash --quiet | ipfs cid format -f '%M' -b base58btc
+QmSju6XPmYLG611rmK7rEeCMFVuL6EHpqyvmEU6oGx3GR8
+```
+
+The rule `//QmSju6XPmYLG611rmK7rEeCMFVuL6EHpqyvmEU6oGx3GR8` will block `/ipfs/bafybeihrw75yfhdx5qsqgesdnxejtjybscwuclpusvxkuttep6h7pkgmze/my/path`, with `QmSju6XPmYLG611rmK7rEeCMFVuL6EHpqyvmEU6oGx3GR8` being the base58-encoded multihash contained in `bafybeihrw75yfhdx5qsqgesdnxejtjybscwuclpusvxkuttep6h7pkgmze`.
+
+We can convert any CID to its multihash with:
+
+```
+$ ipfs cid format -f '%M' -b base58btc bafybeihrw75yfhdx5qsqgesdnxejtjybscwuclpusvxkuttep6h7pkgmze
+QmecDgNqCRirkc3Cjz9eoRBNwXGckJ9WvTdmY16HP88768
+```
+
+Blocking layer recommendation: NameSystem + PathResolver + BlockService.
+
+In order to check for a matching rule, the PathResolver should:
+
+- IPFS path: convert the CID to v1base32 and hash `CIDV1BASE32/PATH` with the
+  hashing functions used in the denylist. Match against declared double-hashes.
+- IPFS path: convert the CID to CIDv0 and hash `CIDV0/PATH` without trailing `/` with the hashing functions used in the denylist. Match against declared double-hashes.
+- IPNS path:
+
+The NameSystem should:
+
+- If NAME is a domain name: Hash `/ipns/NAME` with the hashing functions used in the denylist. Match against declared double-hashes.
+- If NAME is a CID, extract the multihash, encoded with baseb58btc and hash it with the hashing functions used in the denylist. Match against declared double-hashes.
+
+The BlockService should:
+
+- Convert the CID to `CIDV1BASE32/` (keeping the CID codec and adding a slash at the end) and hash it with the hashing functions used in the denylist. Match against declared double-hashes.
+
+- Convert the CID to b58-encoded-multihash (that is CIDv0) and hash the CID string.
+
+##### `/mime/MIMETYPE` `/mime/*`
+
+Blocks content detected to be of the given type. `/mime/*` blocks all the mimetypes and is meant to work with allow rules (all mimetypes blocked except specific ones).
+
+Blocking layer recommendation: Unixfs
+
+Our recommendation is that /mime/ rules automatically set IPFS clients into a
+"unixfs only" mode where only unixfs (+raw blocks) are allowed at the
+BlockService layer, and content type is checked at the Unixfs layer, as the
+blocks get assembled into an actual files. That should cover gateway usage.
+
+#### Allow (or negated) rules
+
+Block items can be prepended by `+`, that items matching the rule are to be allowed
+rather than blocked.
```

Comment:
Maybe `!` for negation, like a .gitignore

https://git-scm.com/docs/gitignore#_pattern_format

Reply:
In all the examples, negations follow supersets of the negated content without a blank line in between.  this is probably an unintended coincidence, but given that `\n-` and `\n+` are used in some syntaxes like markdown to indent list items or derivative/dependent items, this choice of negation syntax *could* confuse some readers who interpret/assume a rule that you can't negate something unless its a prior entry or subset thereof.

Reply:
I'm fine changing the negation operator to `!`, (or supporting them all), but then it should be changed across the doc and not in one place.

Reply:
I have switched to "!"  now.